avogadro/qtplugins/customelements/customelements.cpp

```diff
@@ -86,9 +99,123 @@ void CustomElements::reassign()
+
+  // Prepare the background thread to read in the selected file.
+  if (!m_fileReadThread)
+    m_fileReadThread = new QThread(qobject_cast<QWidget*>(parent()));
+  if (m_threadedReader)
+    m_threadedReader->deleteLater();
+  m_threadedReader = new BackgroundFileFormat(reader);
+  if (m_fileReadMolecule)
+    m_fileReadMolecule->deleteLater();
+  m_fileReadMolecule = new Molecule(qobject_cast<QWidget*>(parent()));
+  m_fileReadMolecule->setData("fileName", fileName.toLocal8Bit().data());
+  m_threadedReader->moveToThread(m_fileReadThread);
+  m_threadedReader->setMolecule(m_fileReadMolecule);
+  m_threadedReader->setFileName(fileName);
+
+  // Setup a progress dialog in case file loading is slow
+  m_progressDialog = new QProgressDialog(qobject_cast<QWidget*>(parent()));
+  m_progressDialog->setRange(0, 0);
+  m_progressDialog->setValue(0);
+  m_progressDialog->setMinimumDuration(750);
+  m_progressDialog->setWindowTitle(tr("Reading File"));
+  m_progressDialog->setLabelText(
+    tr("Opening file '%1'\nwith '%2'").arg(fileName).arg(ident));
+  /// @todo Add API to abort file ops
+  m_progressDialog->setCancelButton(nullptr);
+  connect(m_fileReadThread, SIGNAL(started()), m_threadedReader, SLOT(read()));
+  connect(m_threadedReader, SIGNAL(finished()), m_fileReadThread, SLOT(quit()));
+  connect(m_threadedReader, SIGNAL(finished()),
+          SLOT(backgroundReaderFinished()));
+
+  // Start the file operation
+  m_fileReadThread->start();
+  m_progressDialog->show();
+
+  return true;
+}
+
+void CustomElements::backgroundReaderFinished()
+{
+  QString fileName = m_threadedReader->fileName();
+  if (m_progressDialog->wasCanceled()) {
+    delete m_fileReadMolecule;
+  } else if (m_threadedReader->success()) {
+    if (!fileName.isEmpty()) {
+      m_fileReadMolecule->setData("fileName", fileName.toLocal8Bit().data());
+    } else {
+      m_fileReadMolecule->setData("fileName", Core::Variant());
+    }
+    setMapFromMolecule(m_fileReadMolecule);
+  } else {
+    QMessageBox::critical(qobject_cast<QWidget*>(parent()), tr("File error"),
+                          tr("Error while reading file '%1':\n%2")
+                            .arg(fileName)
+                            .arg(m_threadedReader->error()));
+    delete m_fileReadMolecule;
+  }
+  m_fileReadThread->deleteLater();
+  m_fileReadThread = nullptr;
+  m_threadedReader->deleteLater();
+  m_threadedReader = nullptr;
+  m_fileReadMolecule = nullptr;
+  m_progressDialog->hide();
+  m_progressDialog->deleteLater();
+  m_progressDialog = nullptr;
+}
+
+void CustomElements::setMapFromMolecule(QtGui::Molecule* mol)
+{
+  if (mol->atomCount() != m_molecule->atomCount()) {
+    QMessageBox::critical(
+      qobject_cast<QWidget*>(parent()), tr("Error"),
+      tr("Atom count mismatch.\nExpected %1 atoms, found %2.")
+        .arg(m_molecule->atomCount())
+        .arg(mol->atomCount()));
+  } else {
+    size_t n = m_molecule->atomCount();
```

Comment:
If you get a file, do you want to map bonds as well? For the molecular-based trajectories (even XYZ) this seems like a good idea.